packages/core/useSwipe/index.ts

```diff
@@ -0,0 +1,119 @@
+import { computed, onBeforeMount, onBeforeUnmount, ref, unref, watch } from 'vue-demi'
+import { noop } from './../../shared/utils/is'
+
+import { MaybeRef } from '..'
+
+export enum SwipeDirection {
+  UP = 'UP', RIGHT = 'RIGHT', DOWN = 'DOWN', LEFT = 'LEFT'
+}
+
+/**
+ * Reactive swipe detection.
+ *
+ * @see {@link https://vueuse.js.org/useSwipe}
+ * @param target
+ * @param options
+ */
+export function useSwipe(
+  target: MaybeRef<HTMLElement | null | undefined>,
+  options?: {
+    threshold?: number
+    onSwipe?: (e: TouchEvent) => void
+    onSwipeEnd?: (e: TouchEvent, direction: SwipeDirection) => void
+  },
+) {
+  const targetRef = ref(target)
+  const { threshold, onSwipe, onSwipeEnd } = options || {}
+
+  const coordsStart = ref([0, 0])
+  const coordsEnd = ref([0, 0])
+
+  const diffX = computed(() => coordsStart.value[0] - coordsEnd.value[0])
+  const diffY = computed(() => coordsStart.value[1] - coordsEnd.value[1])
+
+  const isThresholdExeeded = computed(() => Math.max(Math.abs(diffX.value), Math.abs(diffY.value)) >= (threshold || 0))
+
+  const isSwiping = ref(false)
+
+  const direction = computed(() => {
+    if (isThresholdExeeded.value) {
+      if (Math.abs(diffX.value) > Math.abs(diffY.value)) {
+        if (diffX.value > 0) return SwipeDirection.LEFT
+        return SwipeDirection.RIGHT
+      }
+      else {
+        if (diffY.value > 0) return SwipeDirection.UP
+        else return SwipeDirection.DOWN
+      }
+    }
+    else {
+      return null
+    }
+  })
+
+  const getTouchEventCoords = (e: TouchEvent) => [e.touches[0].clientX, e.touches[0].clientY]
+
+  const onTouchStart = (e: TouchEvent) => {
+    coordsStart.value = getTouchEventCoords(e)
+    coordsEnd.value = coordsStart.value
+  }
+
+  const onTouchMove = (e: TouchEvent) => {
+    coordsEnd.value = getTouchEventCoords(e)
+    if (direction.value) {
+      isSwiping.value = true
+      onSwipe?.(e)
+    }
+    else { e.preventDefault() }
+  }
+
+  const onTouchEnd = (e: TouchEvent) => {
+    isSwiping.value = false
+    if (direction.value)
+      onSwipeEnd?.(e, direction.value)
+  }
+
+  const addListeners = (el: HTMLElement | undefined | null) => {
```

Comment:
You can use `useEventListener` which will auto unregister on unmount and target changes.

Reply:
beware using passive event handlers at least on `touchstart`, `touchmove` and  `touchend`: on chromium based it not using passive event handler, swipe effect will freeze the window a few miliseconds

Reply:
you can see an examle on: https://github.com/motss/app-datepicker/blob/main/src/tracker.ts

Reply:
I've now added `preventScrolling` to the composable options. The composable should also use `passive` now, as far the `preventScrolling` is `false`.

Reply:
why not using a simplest way to detect `passive` event handler support? the function is redundant, can be inlined and using typescript definition, you dont need to defineProperty:
```typescript
let supportsPassive = false;
const fn = () => { /* empty listener */ };
const optionsBlock: AddEventListenerOptions = {
  get passive() {
    supportsPassive = true;
    return false;
  }
};
document.addEventListener('x', fn, optionsBlock);
document.removeEventListener('x', fn);
```

Reply:
protect it when using SSR checking for document...

Reply:
Nice, thanks for sharing :)